src/basic/cgroup-util.c

```diff
@@ -312,6 +317,30 @@ int cg_kill(
         return ret;
 }
 
+int cg_kill(
+                const char *controller,
+                const char *path,
+                int sig,
+                CGroupFlags flags,
+                Set *s,
+                cg_kill_log_func_t log_kill,
+                void *userdata) {
+        int r;
+
+        r = cg_kill_items(controller, path, sig, flags, s, log_kill, userdata, "cgroup.procs");
+        if (r < 0)
+                return r;
+
+        r = cg_unified_controller(controller);
+        if (r < 0)
+                return r;
+        if (r == 0) /* doesn't apply to legacy hierarchy */
+                return 0;
+
+        /* Kill remaining threads as a workaround for kernel bug for cgroupsv2, fixed in 5.2-rc5 (c03cd7738a83) */
+        return cg_kill_items(controller, path, sig, flags, s, log_kill, userdata, "cgroup.threads");
```

Comment:
hmm, we probably can further conditionalize this, to only apply to `sig == SIGKILL`, right?

Reply:
It seems that the zombie reacts to other signals too and then for example `unit_kill_context()` can send a SIGHUP. I don't think we can make this optimization.

Reply:
not sure i follow? i mean there are two kinds of signals in a way:

1. SIGKILL + SIGSTOP which are kernel operations on tasks pretty much, which cannot be caught, blocked or ignored. They execute a kernel operation on a task, and userspace cannot  interfere, see signal(7) about that.

2. All the others are mostly just friendly requests to the app: the app can catch them, can block them, can ignore them. The kernel only operates on them if the app doesn't handle them themselves.

This means that there's no need to deliver SIGTERM, SIGHUP and friends to all threads/zombie threads or so of a process, because they are just friendly anyway, and if a process is in such a weird state it's not going to react to friendly requests anyway. OTOH SIGKILL/SIGSTOP is something we need to rely on, because that is supposed to really make evreything go away. Hence as I understand it should matter to deliver SIGKILL (and maybe SIGSTOP) to all those weird threads, but for the other kind of signals we don't have to and probably really shouldn't.

Do I make any sense?

Reply:
OK, I'll make an update.